### PR TITLE
Verify State Root in Block Processing

### LINKED ISSUE #1545: Verify State Root in Block Processing
Hi all, we're missing a simple line:
```
Verify block.state_root == hash_tree_root(state) if there exists a block for the slot being processed.
```
This should be done once we have nil pointer support from #1461

### PR DESCRIPTION
**Pull request by skillful-alex**
Resolves #1545

Implemented state root verification from specification:
`Verify block.state_root == hash_tree_root(state) if there exists a block for the slot being processed.
`
https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#state-root-verification

(I closed PR # 1952 due to a bad commits history and re-creation of the git repository)

### CODE REVIEW

**Comment by skillful-alex** [comment]
The pull request is ready for review and merge.
Please help me change the pull request stage and label. Looks like I don't have permissions for this.

File: beacon-chain/blockchain/service_test.go [hunk 1]
```diff
@@ -533,6 +566,63 @@ func TestReceiveBlock_RemovesPendingDeposits(t *testing.T) {
...
+	if err != nil {
+		t.Fatalf("Can't generate genesis state: %v", err)
+	}
+	parentHash, genesisBlock := setupGenesisBlock(t, chainService, beaconState)
+	beaconState.Slot++
+	if err := chainService.beaconDB.UpdateChainHead(genesisBlock, beaconState); err != nil {
+		t.Fatal(err)
+	}
+
+	// good block processing check
```
Comment (terenc3t) [inline beacon-chain/blockchain/service_test.go]: to be consistent with the rest, do you mind separating `good block processing check` and `bad block processing check` to 2 tests? We like having one test to test one thing

Comment (nisdas) [reply beacon-chain/blockchain/service_test.go]: +1

**Comment by prestonvanloon** [comment]
Let's add a feature flag and merge this PR. You can use this commit as an example feature flag addition: https://github.com/prysmaticlabs/prysm/pull/2050/commits/943b61893bceb3a9ae1d8b62c45fd08ee0462b4b

In short, we should be able to check this in so it doesn't rot, but enable it with `--enable-state-root-verification`

**Comment by prestonvanloon** [comment]
Looks safe to merge. Thanks so much @skillful-alex

**Comment by prestonvanloon** [comment]
failing test on merge, i'm looking into it

File: beacon-chain/blockchain/block_processing.go [hunk 2]
```diff
@@ -80,6 +83,20 @@ func (c *ChainService) ReceiveBlock(ctx context.Context, block *pb.BeaconBlock)
 		return beaconState, fmt.Errorf("could not process block deposits, attestations, and other operations: %v", err)
 	}
 
+	// Check state root
```
Comment (prestonvanloon) [inline beacon-chain/blockchain/block_processing.go]: Actually, don't you think this goes in VerifyBlockValidity?

Comment (prestonvanloon) [reply beacon-chain/blockchain/block_processing.go]: It at least needs to go before cleanup block operations.

**Comment by prestonvanloon** [comment]
Actually, I'm having a hard time reconciling a test failure with compute state root verification. 

```
--- FAIL: TestReceiveBlock_OnChainSplit (0.79s)
    block_processing_test.go:457: beacon state root is not equal to block state root: 0xc894c34db79c73f7b4d8d31b59a30b8d43259ac247498bda94b51e2befb84061 != 0x5b0d5f5910281e4155ad399285483fa76f62d2b331bfe8eb07f03250fe4510c7
```

I tried a few solutions to this, but I still couldn't get it to work which makes me nervous that it wouldn't work in production. Can you take a look at this test @skillful-alex ?

File: beacon-chain/blockchain/block_processing.go (hunk 2 above)
Comment (skillful-alex) [reply beacon-chain/blockchain/block_processing.go]: The check moved above to CleanupBlockOperations.

In the specification, this is a separate step that requires a full state calculation, so I don’t think it’s right to include this in VerifyBlockValidity.

**Pull request merged**